### PR TITLE
feat: serialize and hydrate data before setting it into the ydoc

### PR DESCRIPTION
**Pull request by asalem1**
This PR introduces serialization and hydration of data as it flows in and out of chatrooms so that the API doesn't have to do those operations prior to writing data

### CODE REVIEW

File: src/flows/context/flow.current.tsx [hunk 1]
```diff
@@ -69,37 +69,45 @@ export const FlowProvider: FC = ({children}) => {
...
 
   useEffect(() => {
-    if (isFlagEnabled('sharedFlowEditing')) {
+    const doc = yDoc.current
+    if (isFlagEnabled('sharedFlowEditing') && currentID) {
       provider.current = new WebsocketProvider(
         'wss://demos.yjs.dev', // todo(ariel): replace this with an actual API that we setup
-        flowId, // todo(ariel): we might need to confine this to an org, depends on how multi-org plays out
-        yDoc.current
+        currentID, // todo(ariel): we might need to confine this to an org, depends on how multi-org plays out
```
Comment (drdelambre) [inline src/flows/context/flow.current.tsx:76]: re org: nah. notebook id should be unique across orgs

File: src/flows/context/flow.current.tsx [hunk 2]
```diff
@@ -69,37 +69,45 @@ export const FlowProvider: FC = ({children}) => {
...
       provider.current.on('sync', isSynced => {
         if (isSynced) {
-          const localState = yDoc.current.getMap('localState')?.toJSON()
-          if (!localState) {
-            yDoc.current.getMap('localState').set('localState', currentFlow)
+          const {flowUpdateData} = doc.getMap('flowUpdateData')?.toJSON()
+          if (!flowUpdateData && currentFlow) {
+            doc
+              .getMap('flowUpdateData')
+              .set('flowUpdateData', serialize(currentFlow, orgID))
```
Comment (drdelambre) [inline src/flows/context/flow.current.tsx:85]: we shouldn't be filling in the orgID here. and more generally, why do we need it?

File: src/flows/context/flow.current.tsx [hunk 3]
```diff
@@ -69,37 +69,45 @@ export const FlowProvider: FC = ({children}) => {
...
-          return localState
-        })
+    }
+
+    const onUpdate = () => {
+      const {flowUpdateData} = doc.getMap('flowUpdateData').toJSON()
+      const hydrated = hydrate(flowUpdateData?.data)
+
+      setCurrentFlow(prev => {
+        if (btoa(JSON.stringify(prev)) === btoa(JSON.stringify(hydrated))) {
```
Comment (drdelambre) [inline src/flows/context/flow.current.tsx:96]: does this condition ever fire?

File: src/flows/context/flow.current.tsx [hunk 4]
```diff
@@ -69,37 +69,45 @@ export const FlowProvider: FC = ({children}) => {
...
 
+    doc.on('update', onUpdate)
     return () => {
       if (isFlagEnabled('sharedFlowEditing')) {
         disconnectProvider()
       }
+      doc.off('update', onUpdate)
     }
-  }, [flowId])
+  }, [currentID, currentFlow, orgID])
```
Comment (drdelambre) [inline src/flows/context/flow.current.tsx:110]: remove the orgid, and how often does the disconnect / reconnect logic fire when you're editing a flux script?

File: src/flows/context/flow.current.tsx [hunk 5]
```diff
@@ -114,7 +122,9 @@ export const FlowProvider: FC = ({children}) => {
             ...flowCopy,
             ...flowCopy.data.byID[id],
           }
-          yDoc.current.getMap('localState').set('localState', update)
+          yDoc.current
+            .getMap('flowUpdateData')
+            .set('flowUpdateData', serialize(update, orgID))
```
Comment (drdelambre) [inline src/flows/context/flow.current.tsx:127]: oooo this orgid

File: src/flows/context/flow.current.tsx [hunk 6]
```diff
@@ -158,7 +168,9 @@ export const FlowProvider: FC = ({children}) => {
             ...flowCopy,
             ...flowCopy.meta.byID[id],
```
Comment (drdelambre) [inline src/flows/context/flow.current.tsx:169]: idk why i missed this before, but why are we overwriting the data within the meta element, onto the parent container? i think we're just supposed to be writing flowCopy to the ydoc

File: src/flows/context/flow.current.tsx [hunk 7]
```diff
@@ -198,9 +210,15 @@ export const FlowProvider: FC = ({children}) => {
           for (const ni in flow) {
             flowCopy[ni] = flow[ni]
           }
-          yDoc.current.getMap('localState').set('localState', {
-            ...flowCopy,
-          })
+          yDoc.current.getMap('flowUpdateData').set(
+            'flowUpdateData',
+            serialize(
+              {
```
Comment (drdelambre) [inline src/flows/context/flow.current.tsx:216]: no need to clone the object here

File: src/flows/context/flow.current.tsx (hunk 3 above)
Comment (asalem1) [reply src/flows/context/flow.current.tsx:96]: It shouldn’t anymore I don’t think

File: src/flows/context/flow.current.tsx (hunk 2 above)
Comment (asalem1) [reply src/flows/context/flow.current.tsx:85]: The serializer requires it

File: src/flows/context/flow.current.tsx (hunk 1 above)
Comment (asalem1) [reply src/flows/context/flow.current.tsx:76]: Totally

File: src/flows/context/flow.current.tsx (hunk 6 above)
Comment (asalem1) [reply src/flows/context/flow.current.tsx:169]: Could be. This is ephemeral pasta

File: src/flows/context/flow.current.tsx (hunk 7 above)
Comment (asalem1) [reply src/flows/context/flow.current.tsx:216]: True

File: src/flows/context/flow.current.tsx (hunk 4 above)
Comment (asalem1) [reply src/flows/context/flow.current.tsx:110]: It's pretty slow with the `orgID` included, but in both cases there's a bunch of syncing and disconnecting

**Comment by ui-ci-bot-influx** [comment]
Metric | Score | Level
------------ | ------------- | ------------- 
performance | 57 | :red_circle:
accessibility | 97 | :green_circle:
seo | 90 | :green_circle:
pwa | null | :hankey:
best-practices | 93 | :green_circle:


[Compare your scores with master branch](https://influxdb.aws.influxdata.io/orgs/844910ece80be8bc/dashboards/073c4b95f1089000?lower=now%28%29%20-%202d&vars%5Blighthouse_score_branch%5D=3155)

**Review by drdelambre — APPROVED** [review]
make sure we're not killing existing notebooks as we migrate the hydrate / serialize schema, but other than that this looks good

**Pull request merged**